cpp/src/arrow/compute/cast.cc

```diff
@@ -226,19 +254,237 @@ struct CastFunctor<O, I,
 };
 
 // ----------------------------------------------------------------------
+// Dictionary to other things
+
+template <typename IndexType>
+void UnpackFixedSizeBinaryDictionary(FunctionContext* ctx, const Array& indices,
+                                     const FixedSizeBinaryArray& dictionary,
+                                     ArrayData* output) {
+  using index_c_type = typename IndexType::c_type;
+  const uint8_t* valid_bits = indices.null_bitmap_data();
+  INIT_BITSET(valid_bits, indices.offset());
+
+  const index_c_type* in =
+      reinterpret_cast<const index_c_type*>(indices.data()->buffers[1]->data()) +
+      indices.offset();
+  uint8_t* out = output->buffers[1]->mutable_data();
+  int32_t byte_width =
+      static_cast<const FixedSizeBinaryType&>(*output->type).byte_width();
+  for (int64_t i = 0; i < indices.length(); ++i) {
+    if (bitset_valid_bits & (1 << bit_offset_valid_bits)) {
+      const uint8_t* value = dictionary.Value(in[i]);
+      memcpy(out + i * byte_width, value, byte_width);
+    }
+    READ_NEXT_BITSET(valid_bits);
+  }
+}
+
+template <typename T>
+struct CastFunctor<
+    T, DictionaryType,
+    typename std::enable_if<std::is_base_of<FixedSizeBinaryType, T>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options, const Array& input,
+                  ArrayData* output) {
+    const DictionaryArray& dict_array = static_cast<const DictionaryArray&>(input);
+    const DictionaryType& type = static_cast<const DictionaryType&>(*input.type());
+    const DataType& values_type = *type.dictionary()->type();
+    const FixedSizeBinaryArray& dictionary =
+        static_cast<const FixedSizeBinaryArray&>(*type.dictionary());
+
+    // Check if values and output type match
+    bool types_equal;
+    FUNC_RETURN_NOT_OK(TypeEquals(values_type, *output->type, &types_equal));
+    if (!types_equal) {
+      std::stringstream ss;
+      ss << "Dictionay and target type do not match: ";
+      ss << values_type.ToString() << ", " << output->type->ToString();
+      ctx->SetStatus(Status::Invalid(ss.str()));
+      return;
+    }
```

Comment:
I think this should be a DCHECK for now, since this check failing would indicate a programmer error. We will need to come up with a more general purpose typechecking solution as the compute library grows larger and more user-facing

Reply:
This programming error than must be inside of `cast.cc`. Specifying the wrong target type will already trigger an error earlier?